plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/domains/TypingDomain.kt

```diff
@@ -0,0 +1,272 @@
+
+
+const val TYPE_DOMAIN_NAME = "Type"
+
+/**
+ * Viper Code:
+ * ```
+ * domain Type  {
+ *
+ *   unique function IntType(): Type
+ *   unique function BooleanType(): Type
+ *   unique function UnitType(): Type
+ *   unique function NothingType(): Type
+ *   unique function AnyType(): Type
+ *   unique function FunctionType(): Type
+ *
+ *   function NullableType(t: Type): Type
+ *   function isSubtype(a: Type, b: Type): Bool
+ *   function is_nullable_type(t: Type): Bool
+ *
+ *   axiom {
+ *     (forall t: Type :: { NullableType(t) }
+ *       NullableType(t) != IntType())
+ *   }
+ *   axiom {
+ *     (forall t: Type :: { NullableType(t) }
+ *       NullableType(t) != BooleanType())
+ *   }
+ *   axiom {
+ *     (forall t: Type :: { NullableType(t) }
+ *       NullableType(t) != UnitType())
+ *   }
+ *   axiom {
+ *     (forall t: Type :: { NullableType(t) }
+ *       NullableType(t) != NothingType())
+ *   }
+ *   axiom {
+ *     (forall t: Type :: { NullableType(t) }
+ *       NullableType(t) != AnyType())
+ *   }
+ *   axiom {
+ *     (forall t: Type :: { NullableType(t) }
+ *       NullableType(t) != FunctionType())
+ *   }
+ *   // Axioms of the form NullableType(T) != ClassType()
+ *   axiom {
+ *     (forall t: Type, t2: Type ::
+ *       t != t2 ==> NullableType(t) != NullableType(t2))
+ *   }
+ *
+ *   axiom { !is_nullable_type(IntType()) }
+ *   axiom { !is_nullable_type(BooleanType()) }
+ *   axiom { !is_nullable_type(UnitType()) }
+ *   axiom { !is_nullable_type(NothingType()) }
+ *   axiom { !is_nullable_type(AnyType()) }
+ *   axiom { !is_nullable_type(FunctionType()) }
+ *   // Axioms of the form !is_nullable_type(ClassType())
+ *   axiom {
+ *     (forall t: Type ::
+ *       { NullableType(t) }
+ *       is_nullable_type(t))
+ *   }
+ *
+ *   axiom { isSubtype(IntType(), AnyType()) }
+ *   axiom { isSubtype(BooleanType(), AnyType()) }
+ *   axiom { isSubtype(UnitType(), AnyType()) }
+ *   axiom { isSubtype(NothingType(), AnyType()) }
+ *   axiom { isSubtype(AnyType(), AnyType()) }
+ *   axiom { isSubtype(FunctionType(), AnyType()) }
+ *   // Axioms of the form isSubtype(ClassType(), AnyType())
+ *
+ *   axiom {
+ *     (forall t: Type ::isSubtype(t, t))
+ *   }
+ *   axiom {
+ *     (forall t: Type, t2: Type, t3: Type :: { isSubtype(t, t2), isSubtype(t2, t3) }
+ *       isSubtype(t, t2) && isSubtype(t2, t3) ==> isSubtype(t, t3))
+ *   }
+ *   axiom {
+ *     (forall t: Type, t2: Type :: { isSubtype(t, t2), isSubtype(t2, t) }
+ *       isSubtype(t, t2) && isSubtype(t2, t) ==> t == t2)
+ *   }
+ *
+ *   axiom {
+ *     (forall t: Type, t2: Type :: { isSubtype(NullableType(t), NullableType(t2)) }
+ *       isSubtype(t, t2) ==> isSubtype(NullableType(t), NullableType(t2)))
+ *   }
+ *
+ *   axiom {
+ *     (forall t: Type :: { isSubtype(NothingType(), t) }
+ *       isSubtype(NothingType(), t))
+ *   }
+ * }
+ * ```
+ */
+class TypeDomain(classes: List<ClassEmbedding>) : BuiltinDomain(TYPE_DOMAIN_NAME) {
```

Comment:
I think the use of a bare string here is an oversight on my part: `Domain` should take a `MangledName`, so you can do `BuiltinDomain(DomainName("whatever you want"))` and then use `this.name` to refer to it, saving the need for a constant.

Reply:
You can already do that. However, the reason that doesn't work here is because I need the name of the domain in the companion object where I don't have a reference to the `Domain` object.

Reply:
Perhaps this would be better done in `init`?

Reply:
You mean putting the code that is currently in the companion object into an `init`? The problem with that is that the whole reason I put these things into an companion object is that other files should be able to access them without needing a reference to the `TypeDomain` object as a lot of the domain functions are actually static and don't depend on the user defined classes.